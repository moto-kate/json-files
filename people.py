import json
import sys
import schema
from operator import itemgetter
from functools import reduce


def get_cl_params():
    file_name = sys.argv[1]
    command = sys.argv[2]
    return {'file_name': file_name, 'command': command}


def read_file(file_name: str) -> str:
    with open(file_name, 'r') as file:
        file_content = file.read()
        return file_content


def parse_json(json_as_str: str) -> dict:
    return json.loads(json_as_str)


people_schema = schema.Schema(
    {
        'people': [
            {
                'first_name': str,
                'last_name': str,
                'currency': schema.Or('EUR', 'USD', 'PLN'),
                'salary': int,
                'height': int,
            }
        ]
    }
)


def validate_people(people):
    people_schema.validate(people)


def get_people_list(file_name: str) -> list[dict]:
    people_as_json_str = read_file(file_name)
    people = parse_json(people_as_json_str)
    validate_people(people)
    return people.get('people')


def write_file(file_name: str, content: str):
    with open(file_name, 'w') as file:
        file.write(content)


def get_full_name(person):
    first_name, last_name = person.get('first_name'), person.get('last_name')
    if first_name and last_name:
        return first_name + ' ' + last_name
    return first_name + last_name


def display_eur_earners(people: list[dict]):
    people_earning_in_eur = filter(
        lambda person: person.get('currency') == 'EUR', people)
    print('People earning in EUR:')
    for person in people_earning_in_eur:
        print('  ' + get_full_name(person))


def display_avg_height(people: list[dict]):
    total = reduce(lambda sum, person: sum + person.get('height'), people, 0)
    avg_height = total / len(people)
    print('Average height:', '%.2f' % avg_height, 'cm')


def display_stats(people: list[dict]):
    display_eur_earners(people)
    display_avg_height(people)


def read_person_from_input():
    print('Please specify person data:')
    first_name = input('First name? ')
    last_name = input('Last name? ')
    height = int(input('Height [cm]? '))
    currency = input('Currency [EUR, USD, PLN]? ')
    salary = int(input('Salary? '))
    return {
        'first_name': first_name,
        'last_name': last_name,
        'height': height,
        'currency': currency,
        'salary': salary
    }


def add_person(file_name: str, people: list[dict]):
    person = read_person_from_input()
    updated_people = people.copy().append(person)
    write_file(file_name, json.dumps({'people': updated_people}, indent=2))


def modify_person(file_name: str, people: list[dict]):
    last_name = input('Last name? ')
    new_salary = int(input('New salary? '))
    updated_people = list(map(
        lambda person: dict(
            person, **{'salary': new_salary}) if person.get('last_name') == last_name else person,
        people
    ))
    write_file(file_name, json.dumps({'people': updated_people}, indent=2))


def handle_error(message: str, e: Exception):
    print(message)
    print(type(e))
    exit(1)


def main():
    try:
        file_name, command = itemgetter(
            'file_name', 'command')(get_cl_params())
    except Exception as e:
        handle_error('Missing or incorrect command line parameters', e)

    try:
        people_list = get_people_list(file_name)
    except Exception as e:
        handle_error('Missing or corrupt json file', e)

    try:
        if command == '--stats':
            display_stats(people_list)
        elif command == '--add':
            add_person(file_name, people_list)
        elif command == '--mod':
            modify_person(file_name, people_list)
        else:
            print('Unknown command.')
    except Exception as e:
        handle_error('Could not execute command', e)


if __name__ == '__main__':
    main()
