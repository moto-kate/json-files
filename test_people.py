import sys
import pytest
import schema
import json
import people
from approvaltests.approvals import verify
from approvaltests import verify_all_combinations
from people import parse_json, validate_people, get_people_list, \
    get_full_name, display_eur_earners, display_avg_height, \
    read_person_from_input, add_person

test_person_1 = {
    'first_name': 'John',
    'last_name': 'Doe',
    'height': 179,
    'currency': 'EUR',
    'salary': 5100
}
test_person_2 = {
    'first_name': 'Jane',
    'last_name': 'Marsch',
    'height': 165,
    'currency': 'PLN',
    'salary': 6700
}

test_people_list = [test_person_1, test_person_2]

# @pytest.fixture
# def mock_argv():


def test_get_cl_params(monkeypatch):
    file_name = 'file.json'
    command = '--stats'
    monkeypatch.setattr(sys, 'argv', ['', file_name, command])

    from people import get_cl_params

    assert get_cl_params() == {
        'file_name': file_name, 'command': command}

# def test_read_file(mocker):


def test_parse_json_ok():
    assert parse_json('{ "a": 1 }') == {'a': 1}


def test_parse_json_fail():
    with pytest.raises(json.decoder.JSONDecodeError):
        parse_json('{ "a: 1 }')


def test_validate_people_ok():
    validate_people({'people': [test_person_1, test_person_2]})


def test_validate_people_fail():
    with pytest.raises(schema.SchemaError):
        validate_people({'people': [{}]})


def test_get_people_list(mocker):
    test_file_name = 'test-filename.json'
    test_people = {'people': [test_person_1]}
    mocker.patch('people.read_file', return_value='')
    mocker.patch('people.parse_json', return_value=test_people)
    mocker.patch('people.validate_people', return_value=True)
    read_file_spy = mocker.spy(people, 'read_file')
    parse_json_spy = mocker.spy(people, 'parse_json')
    validate_people_spy = mocker.spy(people, 'validate_people')

    assert get_people_list(test_file_name) == [test_person_1]
    read_file_spy.assert_called_once_with(test_file_name)
    parse_json_spy.assert_called_once_with('')
    validate_people_spy.assert_called_once_with(test_people)


# @pytest.mark.skip('WIP')
# def test_write_file():


@pytest.mark.parametrize(
    'person,full_name', [
        ({'first_name': 'John', 'last_name': 'Rambo'}, 'John Rambo'),
        ({'first_name': '', 'last_name': ''}, '')
    ]
)
def test_get_full_name(person, full_name):
    assert get_full_name(person) == full_name


def test_display_eur_earners(capsys):
    display_eur_earners(test_people_list)

    captured_output = capsys.readouterr().out
    assert captured_output == 'People earning in EUR:\n  John Doe\n'


def test_display_avg_height(capsys):
    display_avg_height(test_people_list)

    captured_output = capsys.readouterr().out
    assert captured_output == 'Average height: 172.00 cm\n'


# def test_display_stats():

def test_read_person_from_input(monkeypatch):
    def mock_inputs():
        yield 'Logan'
        yield 'Wolverine'
        yield '210'
        yield 'USD'
        yield '21000'

    input_values_generator = mock_inputs()
    monkeypatch.setattr(
        'builtins.input', lambda _: next(input_values_generator))

    # Instead of:
    # assert read_person_from_input() == {
    #     'first_name': 'Logan',
    #     'last_name': 'Wolverine',
    #     'height': 210,
    #     'currency': 'USD',
    #     'salary': 21000
    # }
    # verify(read_person_from_input())
    verify_all_combinations(read_person_from_input, ())
    # ( (some args,...), (other args...) )
    # https://snyk.io/advisor/python/approvaltests/functions/approvaltests.verify_all_combinations


def test_add_person(mocker):
    people_list = [test_person_1]
    file_name = 'f.json'

    mocker.patch('people.read_person_from_input', return_value=test_person_2)
    mocker.patch('people.write_file', return_value=None)
    write_file_spy = mocker.spy(people, 'write_file')

    add_person(file_name, people_list)

    write_file_spy.assert_called_once_with(
        file_name, '{\n  "people": null\n}')
