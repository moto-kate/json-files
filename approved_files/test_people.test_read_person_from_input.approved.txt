args: () => {'first_name': 'Logan', 'last_name': 'Wolverine', 'height': 210, 'currency': 'USD', 'salary': 21000}
